File: packages/gavio-js/src/interceptors/reliability/load-balancer.ts
/** loadBalancer (F-REL-04) — weighted round-robin across provider adapters. */

import type { InterceptorContext } from '../../context.js'
import type { ProviderAdapter } from '../../providers/base.js'
import type { GavioRequest } from '../../request.js'
import type { GavioResponse } from '../../response.js'
import { coerceProvider } from '../../types.js'
import type { Executor, ExecutorPolicy } from '../base.js'

export interface LoadBalancerOptions {
  weights?: number[]
}

class LoadBalancer implements ExecutorPolicy {
  readonly name = 'load_balancer'
  readonly isExecutorPolicy = true as const

  private readonly pool: ProviderAdapter[]
  private index = 0

  constructor(adapters: ProviderAdapter[], options: LoadBalancerOptions = {}) {
    if (adapters.length === 0) {
      throw new Error('loadBalancer requires at least one adapter')
    }
    const weights = options.weights ?? adapters.map(() => 1)
    if (weights.length !== adapters.length) {
      throw new Error('weights must match adapters length')
    }
    // Expand by weight, then cycle for round-robin.
    this.pool = []
    adapters.forEach((adapter, i) => {
      for (let k = 0; k < Math.max(1, weights[i]!); k++) this.pool.push(adapter)
    })
  }

  async around(
    request: GavioRequest,
    ctx: InterceptorContext,
    _callNext: Executor,
  ): Promise<GavioResponse> {
    ctx.markFired(this.name)
    const adapter = this.pool[this.index % this.pool.length]!
    this.index += 1
    const rerouted = request.copyWithMessages(request.messages)
    rerouted.provider = coerceProvider(adapter.providerName)
    return adapter.complete(rerouted)
  }
}

/** Factory: build a load balancer over a pool of adapters. */
export function loadBalancer(
  adapters: ProviderAdapter[],
  options: LoadBalancerOptions = {},
): ExecutorPolicy {
  return new LoadBalancer(adapters, options)
}

File: packages/gavio-py/gavio/interceptors/reliability/load_balancer.py
"""LoadBalancer (F-REL-04) — distribute calls across provider adapters."""

from __future__ import annotations

import itertools
import logging

from ...context import InterceptorContext
from ...interceptors.chain import Executor
from ...request import GavioRequest
from ...response import GavioResponse
from ...types import Provider
from ..executor import ExecutorPolicy

logger = logging.getLogger("gavio.loadbalancer")


class LoadBalancer(ExecutorPolicy):
    """Weighted round-robin across a pool of provider adapters.

    Register via ``.use(LoadBalancer([...]))``; it replaces the gateway's single
    provider for the call, routing each request to the next adapter in the
    weighted rotation. ``weights`` (parallel to ``adapters``) bias the rotation.
    """

    def __init__(
        self,
        adapters: list,
        weights: list[int] | None = None,
    ) -> None:
        if not adapters:
            raise ValueError("LoadBalancer requires at least one adapter")
        if weights is not None and len(weights) != len(adapters):
            raise ValueError("weights must match adapters length")
        weights = weights or [1] * len(adapters)
        # Expand by weight, then cycle for round-robin.
        expanded: list = []
        for adapter, weight in zip(adapters, weights, strict=True):
            expanded.extend([adapter] * max(1, weight))
        self._adapters = adapters
        self._cycle = itertools.cycle(expanded)

    @property
    def name(self) -> str:
        return "load_balancer"

    async def around(
        self,
        request: GavioRequest,
        ctx: InterceptorContext,
        call_next: Executor,
    ) -> GavioResponse:
        ctx.mark_fired(self.name)
        adapter = next(self._cycle)
        logger.debug("load balancing to %s", adapter.provider_name)
        rerouted = request.copy_with_messages(request.messages)
        rerouted.provider = Provider.coerce(adapter.provider_name)
        return await adapter.complete(rerouted)
